### ISSUE TITLE
Python local variables are leaked when calling an overloaded .NET method inside an `except` block

### ISSUE THREAD

**Issue by mhsmith** [pythonnet/pythonnet#2279]
### Environment

-   Pythonnet version: 3.0.3
-   Python version: Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)] on win32
-   Operating System: Windows 10
-   .NET Runtime: .NET Framework 4.8.9181.0

### Details
```py
import gc
import sys

import clr
from System import Console

def get_referrers(obj):
    return [x for x in gc.get_referrers(obj) if x is not globals()]

class C:
    pass

c = C()

def f(c):
    try:
        raise Exception("boom")
    except Exception:
        Console.WriteLine("Hello")

print(sys.getrefcount(c))
f(c)
gc.collect()
print(sys.getrefcount(c))
print(get_referrers(c))

frame = get_referrers(c)[0]
print(get_referrers(frame))

tb = get_referrers(frame)[0]
print(get_referrers(tb))

exc = get_referrers(tb)[0]
print(get_referrers(exc))

exc1 = get_referrers(exc)[0]
print(get_referrers(exc1))

import traceback
traceback.print_exception(exc1)
```

Output:
```
2
Hello
3
[<frame at 0x000001B570A0AF00, file 'C:\\Users\\smith\\cygwin\\tmp\\leak.py', line 25, code f>]
[<traceback object at 0x000001B570FAE600>]
[Exception('boom')]
[TypeError('an integer is required'), TypeError('an integer is required'), TypeError('an integer is required'), TypeError('an integer is required'), TypeError("'str' value cannot be converted to System.Char"), TypeError("'str' value cannot be converted to System.Single"), TypeError("'str' value cannot be converted to System.Double"), TypeError("'str' value cannot be converted to System.Boolean"), TypeError("'str' value cannot be converted to System.Decimal")]
[]
Traceback (most recent call last):
  File "C:\Users\smith\cygwin\tmp\leak.py", line 23, in f
    raise Exception("boom")
Exception: boom

During handling of the above exception, another exception occurred:

TypeError: an integer is required
```

The 2 is expected because, as the getrefcount documentation says, "it includes the (temporary) reference as an argument to [getrefcount](https://docs.python.org/3/library/sys.html#sys.getrefcount)". But the 3 indicates that the object was leaked during the call of `f`, even though `f` doesn't pass it to .NET, or use it in any other way.

The object is acquiring an extra reference through the following chain:
* `c` is a local variable within the function `f`.
* The stack frame of `f` is being kept alive because it's inside the traceback of the `Exception`.
* The `Exception` is being kept alive by some `TypeErrors` which were apparently thrown and caught internally by Python.NET while resolving the overload of `WriteLine`. These `TypeErrors` don't have any tracebacks of their own, but they point to the `Exception` as their `__context__`.
* And that's as far as `get_referrers` can take us.

**Comment by mhsmith** [comment]
Now that I think about it, maybe case 1 is working as expected, because .NET's own garbage collector needs to run to clean up the reference to the Python object. But case 2 looks like something different.

**Comment by lostmsu** [comment]
Unless you do this in a loop and memory steadily grows while .NET GC is doing its work, this is not a valid memory leak bug for the Python.NET

**Comment by mhsmith** [comment]
I can't get the memory usage to grow steadily -- it appears that the issue only reliably affects the objects that were on the stack the most recent time this happened. 

However, I think this is still a bug, because any object that appears in the same stack frame as a .NET method call under these circumstances will be kept alive until the next time the circumstances occur, which may be never.

Repeatedly running both the Python and .NET garbage collectors doesn't help. For example, here we use weakrefs to detect when objects are garbage collected:
```py
import gc
import os
import weakref
from time import sleep

import clr
from System import Console, GC

class C:
    def __init__(self):
        self.big = os.urandom(100 * 1024 * 1024)

def f(x):
    try:
        raise Exception("boom")
    except Exception:
        Console.WriteLine("Hello")

obj_1 = C()
obj_2 = C()
wr_1 = weakref.ref(obj_1)
wr_2 = weakref.ref(obj_2)

f(obj_1)
del obj_1
del obj_2

while True:
    gc.collect()
    GC.Collect()
    sleep(0.5)
    print(bool(wr_1()), bool(wr_2()))
```

Output:
```
Hello
True False
True False
True False
... continues for at least 30 seconds
```

In other words, calling `f(obj_1)` caused `obj_1` to be kept alive, even though `f` didn't do anything with it. If we call `f(obj_2)` instead, then the output changes to `False True`. 

And if we comment out the `WriteLine` call, then we get `False False`.

**Comment by lostmsu** [comment]
> any object that appears in the same stack frame as a .NET method call under these circumstances will be kept alive until the next time the circumstances occur, which may be never

This is the best guarantee that .NET runtime provides by design.

**Comment by mhsmith** [comment]
The object in question is not a .NET object, nor has it ever been passed to .NET in any way. So it seems unreasonable that its lifetime should become subject to the .NET garbage collector.

It's well-known that maintaining a reference to a Python exception object will cause local variables on the stack to be kept alive. And I think I've provided clear evidence that Python.NET is holding persistent references to, in this case, as many as 9 Python exceptions which should already have been caught and disposed of. 

Even if this can't be turned into an unbounded memory leak, I think it's still a bug.
